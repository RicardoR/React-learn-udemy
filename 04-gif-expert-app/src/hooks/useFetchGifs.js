import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = (category) => {
	const [state, setState] = useState({
		data: [],
		loading: true,
	});

	// useEffect only triggers when categroy changes
	useEffect(() => {
		getGifs(category).then((imgs) => {
			setTimeout(() => {
				setState({
					data: imgs,
					loading: false,
				});
			}, 1000);
		});
  }, [category]);

  return state;
};