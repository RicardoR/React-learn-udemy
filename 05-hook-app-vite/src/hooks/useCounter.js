import { useState } from 'react';

export const useCounter = (inititalValue = 10) => {
  const [counter, setCounter] = useState(inititalValue);

  const reset = () => {
    setCounter(inititalValue);
  };

  const increment = (value = 1) => {
    setCounter(counter + value);
  };

  const decrement = (value = 1) => {
    if (counter - value <= 0) {
      return;
    }
    setCounter(counter - 1);
  };
  return {
    counter,
    increment,
    decrement,
    reset,
  };
};
