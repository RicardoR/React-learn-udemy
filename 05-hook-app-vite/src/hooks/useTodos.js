import {useEffect, useReducer, useState} from 'react';
import {addTodoAction, deleteTodoAction, todoReducer, toggleTodoAction} from "../08-useReducer/todoReducer";

const LOCALSTORAGE_TODOS_KEY = 'todos';
const initialState = []

const init = () => {
    return JSON.parse(localStorage.getItem(LOCALSTORAGE_TODOS_KEY)) ?? [];
}

export const useTodos = () => {

    const [todos, dispatch] = useReducer(todoReducer, initialState, init);

    const [pendingTodosCount, setPendingTodosCount] = useState(0);
    const [todosCount, setTodosCount] = useState(0);

    useEffect(() => {
        const todosToStore = JSON.stringify(todos);
        localStorage.setItem(LOCALSTORAGE_TODOS_KEY, todosToStore);
        const pending = todos.filter(todo => todo.done === false)?.length;
        setPendingTodosCount(pending);
        setTodosCount(todos.length);
    }, [todos]);

    const handleNewTodo = (todo) => {
        const action = {
            type: addTodoAction,
            payload: todo
        }
        dispatch(action);
    }

    const handleDeleteTodo = (todo) => {
        const action = {
            type: deleteTodoAction,
            payload: todo
        };
        dispatch(action);
    }

    const handleToggleTodo = (todo) => {
        const action = {
            type: toggleTodoAction,
            payload: todo
        };
        dispatch(action);
    }

    return {
        todos,
        todosCount,
        pendingTodosCount,
        handleNewTodo,
        handleDeleteTodo,
        handleToggleTodo
    }
};

