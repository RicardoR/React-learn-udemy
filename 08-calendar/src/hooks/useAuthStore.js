import { useDispatch, useSelector } from 'react-redux';
import calendarApi from '../api/calendarApi';
import { clearErrorMessage, onChecking, onLogin, onLogout } from '../store';

// Interact with auth store

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  const startLogin = async ({ email, password }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post('/auth', { email, password });
      storeTokenInLocalStorage(data.token);
      dispatchLogin(data);
    } catch (error) {
      dispatch(onLogout('Wrong credentials'));
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  const startRegister = async ({ name, email, password }) => {
    dispatch(onChecking());

    try {
      const { data } = await calendarApi.post('/auth/new', {
        name,
        email,
        password,
      });
      storeTokenInLocalStorage(data.token);
      dispatchLogin(data);
    } catch (error) {
      const err = error.data?.msg ?? 'Error';
      dispatch(onLogout(err));

      console.log(error);
      console.log(error.response.data);
    }
  };

  const startLogout = () => {
    localStorage.clear();
    dispatch(onLogout());
  };

  const checkAuthToken = async () => {
    const token = localStorage.getItem('token');

    if (!token) {
      return dispatch(onLogout());
    }

    try {
      const { data } = await calendarApi.get('/auth/renew');
      storeTokenInLocalStorage(data.token);
      dispatchLogin(data);
    } catch (error) {
      startLogout();
    }
  };

  const storeTokenInLocalStorage = (token) => {
    localStorage.setItem('token', token);
    localStorage.setItem('token-init-date', new Date().getTime());
  };

  const dispatchLogin = (data) => {
    dispatch(onLogin({ name: data.name, uid: data.uid }));
  };

  return {
    // Props
    status,
    user,
    errorMessage,
    // Methods
    startLogin,
    startRegister,
    startLogout,
    checkAuthToken,
  };
};
