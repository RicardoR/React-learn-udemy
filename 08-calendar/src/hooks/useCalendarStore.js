import { useDispatch, useSelector } from 'react-redux';
import Swal from 'sweetalert2';
import calendarApi from '../api/calendarApi';
import { convertEventsToDateEvents } from '../helpers';
import {
  onSetActiveEvent,
  onAddNewEvent,
  onUpdateEvent,
  onDeleteEvent,
  onLoadEvents,
} from '../store';

export const useCalendarStore = () => {
  const dispatch = useDispatch();
  const { events, activeEvent } = useSelector((state) => state.calendar);
  const { user } = useSelector((state) => state.auth);

  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  const startLoadingEvents = async () => {
    try {
      const { data } = await calendarApi.get('/events');
      const eventsConverted = convertEventsToDateEvents(data.events);
      dispatch(onLoadEvents(eventsConverted));
    } catch (error) {
      console.log('Error getting events', error);
    }
  };

  const startSavingEvent = async (calendarEvent) => {
    try {
      if (calendarEvent.id) {
        await calendarApi.put(`/events/${calendarEvent.id}`, calendarEvent);
        dispatch(onUpdateEvent({ ...calendarEvent, user }));
        return;
      }

      const { data } = await calendarApi.post('/events', calendarEvent);
      dispatch(
        onAddNewEvent({
          ...calendarEvent,
          id: data.evento.id,
          user: user,
        })
      );
    } catch (error) {
      Swal.fire('Error saving event', error.response?.data?.msg, 'error');
    }
  };

  const startDeletingEvent = () => {
    dispatch(onDeleteEvent());
  };

  return {
    // Props
    events,
    activeEvent,
    hasEventSelected: !!activeEvent,

    // Methods
    setActiveEvent,
    startSavingEvent,
    startDeletingEvent,
    startLoadingEvents,
  };
};
