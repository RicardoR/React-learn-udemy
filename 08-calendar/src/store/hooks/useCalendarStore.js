import { useDispatch, useSelector } from 'react-redux';
import { onSetActiveEvent, onAddNewEvent, onUpdateEvent } from '../';

export const useCalendarStore = () => {
  const dispatch = useDispatch();
  const { events, activeEvent } = useSelector((state) => state.calendar);

  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  const startSavingEvent = async (calendarEvent) => {
    // todo: call back
    if (calendarEvent._id) {
      dispatch(onUpdateEvent({ ...calendarEvent }));
    } else {
      dispatch(
        onAddNewEvent({
          ...calendarEvent,
          _id: new Date().getTime(),
        })
      );
    }
  };

  return {
    // Props
    events,
    activeEvent,

    // Methods
    setActiveEvent,
    startSavingEvent,
  };
};
