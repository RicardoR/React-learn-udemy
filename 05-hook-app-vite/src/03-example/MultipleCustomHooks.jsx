import {useCounter, useFetch} from "../hooks";
import {LoadingQuote, Quote} from "./";


export const MultipleCustomHooks = () => {

    const {counter, increment} = useCounter(1)
    const {data, isLoading, hasError} = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
    const {author, quote} = !!data && data[0];

    return (
        <>
            <h1> breaking bad Quotes</h1>
            <hr/>

            {
                isLoading
                    ? <LoadingQuote></LoadingQuote>
                    : <Quote quote={quote} author={author}></Quote>
            }

            <button className='btn btn-primary'
                    disabled={isLoading}
                    onClick={() => increment()}>
                Next quote
            </button>
        </>
    );
};



