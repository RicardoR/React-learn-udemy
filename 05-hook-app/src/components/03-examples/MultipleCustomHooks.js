import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';
import './multiple-custom.css';

export const MultipleCustomHooks = () => {
  const { counter, increment } = useCounter(1);

  const { data, loading } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );

  const { author, quote } = !!data && data[0];

  return (
    <>
      <h1>Breaking bad quotes</h1>
      <hr />
      {loading ? (
        <div className="alert alert-info text-center">Loading...</div>
      ) : (
        <blockquote className="blockquote text-right">
          <p className="mb-2">{quote}</p>
          <footer className="blockquote-footer"> {author}</footer>
        </blockquote>
      )}

      <button className="btn btn-primary" onClick={() => increment(1)}>
        Siguiente frase
      </button>
    </>
  );
};
