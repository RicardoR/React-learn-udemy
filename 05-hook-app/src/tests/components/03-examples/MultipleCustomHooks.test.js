import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

// esto dice que si se usa el fichero que contiene el useFetch usa el mock
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('MultipleCustomHooks test', () => {
  beforeEach(() => {
    useCounter.mockReturnValue({
      counter: 10,
      increment: jest.fn(),
    });
  });

  test('should render correctly', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should display the info', () => {
    useFetch.mockReturnValue({
      data: [
        {
          id: '123',
          author: 'Javier',
          quote: 'Hola',
        },
      ],
      loading: false,
      error: null,
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper.find('.alert').exists()).toBe(false);
    expect(wrapper.find('.mb-2').text().trim()).toBe('Hola');
    expect(wrapper.find('footer').text().trim()).toBe('Javier');
    expect(wrapper).toMatchSnapshot();
  });
});
