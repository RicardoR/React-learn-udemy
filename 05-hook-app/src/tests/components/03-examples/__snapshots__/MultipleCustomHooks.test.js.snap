// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleCustomHooks test should display the info 1`] = `
<Fragment>
  <h1>
    Breaking bad quotes
  </h1>
  <hr />
  <blockquote
    className="blockquote text-right"
  >
    <p
      className="mb-2"
    >
      Hola
    </p>
    <footer
      className="blockquote-footer"
    >
       
      Javier
    </footer>
  </blockquote>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente frase
  </button>
</Fragment>
`;

exports[`MultipleCustomHooks test should render correctly 1`] = `
<Fragment>
  <h1>
    Breaking bad quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente frase
  </button>
</Fragment>
`;
